### DISCUSSION TITLE
Troubleshooting runner: zombie process

### DISCUSSION THREAD

**Discussion by GreyTeardrop** [runs-on/runs-on#99] [Q&A] [answered]
Hi!

We're facing an odd issue with RunsOn runners at the moment. It might again be an issue on GitHub side, but their status page is green for now.

Some jobs seem to properly allocate a RunsOn runner, however, the runner won't pick up the job. Nothing suspicious in the RunsOn application logs; when SSHing onto the runner instance, there's a zombie process on it. It seems to be a child of the runs-on-agent:
```
$ ps axo stat,ppid,pid,comm
STAT    PPID     PID COMMAND
...
Sl       564     570 runs-on-agent
Z        570     573 bash <defunct>
...
```

Also, I'm not able to get access to cloud-init logs:
```
$ cat /var/log/cloud-init-output.log
cat: /var/log/cloud-init-output.log: Permission denied
```

Oddly enough, the issue seems to be isolated to runners from a single repository, others seem to be ok at the moment. RunsOn GitHub app has access to all repos, and jobs from the affected repos used to be fine on previous runs ~4 days ago.

Looks like we can reproduce the issue by re-running jobs on the affected repo. How can we go with further troubleshooting?

Thank you,
Mykola

**Comment by crohr** [comment]
@GreyTeardrop you probably need to `sudo cat /var/log/cloud-init-output.log` to view the log output. Also make sure you are using the latest version, which automatically sends the output of that log file to CloudWatch

* https://github.com/runs-on/runs-on/releases/tag/v2.3.2
* https://runs-on.com/guides/troubleshoot/#view-the-instance-logs

If you manage to get the output for a hanging runner, you can send it to cyril@runs-on.com

**Reply by GreyTeardrop** [reply]
Thank you, Cyril! I was so sure I got the log without sudo last time that I didn't even try to use one. I got one and sent you the complete log. Last lines of the log don't seem to be much helpful though…
```

√ Connected to GitHub

{"level":"info","time":"2024-06-26T09:07:30Z","message":"Waiting for tags file to be present..."}
Created symlink /etc/systemd/system/multi-user.target.wants/amazon-cloudwatch-agent.service → /etc/systemd/system/amazon-cloudwatch-agent.service.
Current runner version: '2.317.0'
2024-06-26 09:07:31Z: Listening for Jobs
{"level":"info","time":"2024-06-26T09:07:33Z","message":"Waiting for tags file to be present..."}
{"level":"info","time":"2024-06-26T09:07:36Z","message":"Waiting for tags file to be present..."}
{"level":"info","time":"2024-06-26T09:07:39Z","message":"Waiting for tags file to be present..."}
{"level":"info","time":"2024-06-26T09:07:42Z","message":"Waiting for tags file to be present..."}
{"level":"info","time":"2024-06-26T09:07:45Z","message":"Waiting for tags file to be present..."}
{"level":"info","time":"2024-06-26T09:07:48Z","message":"Waiting for tags file to be present..."}
```

**Comment by crohr** [comment, marked as answer]
@GreyTeardrop could you try defining your runs-on syntax with the following format:

Instead of e.g. `runs-on: runs-on,runner=2cpu-linux-x64`, do:

```yaml
runs-on:
  labels:
    - runs-on
    - runner=2cpu-linux-x64
    - run-id=${{ github.run_id }}
```

**Reply by GreyTeardrop** [reply]
Wow, that helps! I thought that is supposed to be equivalent though…
BTW, 
```yaml
runs-on:
  - runs-on
  - runner=2cpu-linux-x64
  - run-id=${{ github.run_id }}
```
also works

**Reply by GreyTeardrop** [reply]
I haven't got any response from GitHub support yet. My guess so far is GitHub changed something on its side and for some reason they's not accepting labels like `runs-on,runner=2cpu-linux-x64` now, right? Perhaps they stumble on the comma symbol? They are probably rolling that update out, so it only affects some repos.

**Reply by crohr** [reply]
Yes that’s what I think as well. I will probably recommend the array syntax from now on in the docs

**Reply by GreyTeardrop** [reply]
I just got a message from GitHub that they've fixed the issue with commas in runner labels. All our repos seem to work with single-label syntax again.

**Reply by crohr** [reply]
Oh interesting, would you mind sharing the reply? Do they confirm it is supposed to be a valid label syntax?

**Reply by GreyTeardrop** [reply]
I missed to explicitly ask that, but that seems to be implied:

> Good news! Our engineering team implemented some changes last night which they believe will address this issue. Please can you confirm you're no longer facing issue with runner assignment while using single-line sytax? We apologize for the lengthly investigation but finding root cause involved troubleshooting across numerous teams at GitHub. We hope the issue should now be resolved for you.

**Reply by crohr** [reply]
Indeed, thanks a lot
